### PR TITLE
[examples][MLIRVector] improve vector dialect examples for memory access

### PR DESCRIPTION
**Pull request by Origami404**
add or improve examples for memory access related operations in vector dialect, including:

- load/store
- expandload/compressstore
- maskedload/maskedstore
- gather/scatter

also do a rename to match operation name better:

- `expand-load` => `expandload`

This PR is a small and finished part of WIP PR #90 .

### CODE REVIEW

File: examples/MLIRVector/vector-expandload.mlir [hunk 1]
```diff
@@ -0,0 +1,102 @@
...
+
+  %base0 = memref.get_global @gv0 : memref<8xi32>
+  %base1 = memref.get_global @gv1 : memref<4x4xi32>
+  %base2 = memref.get_global @gv2 : memref<8xi32>
+
+  %pass_thru_4 = arith.constant dense<[2330, 2331, 2332, 2333]> : vector<4xi32>
+  %pass_thru_8 = arith.constant dense<[2330, 2331, 2332, 2333, 2334, 2335, 2336, 2337]> : vector<8xi32>
+
+  // expandload normal usage
+  // expandload require a pass through value at any time
```
Comment (zhanghb97) [inline examples/MLIRVector/vector-expandload.mlir:36]: require -> requires

File: examples/MLIRVector/vector-expandload.mlir [hunk 2]
```diff
@@ -0,0 +1,102 @@
...
+  %mask0 = arith.constant dense<[1, 0, 1, 0]> : vector<4xi1>
+  
+  // %v0 will be [0, 2331, 1, 2333] instead of [0, 2331, 2, 2333]
+  // because position 1 is masked off, so it will not move to load next element
```
Comment (zhanghb97) [inline examples/MLIRVector/vector-expandload.mlir:40]: Remove "because".

File: examples/MLIRVector/vector-expandload.mlir [hunk 3]
```diff
@@ -0,0 +1,102 @@
...
+  %mask1 = arith.constant dense<[1, 0, 0, 1]> : vector<4xi1>
+
+  %v1 = vector.expandload %base1[%c0, %c0], %mask1, %pass_thru_4
+    : memref<4x4xi32>, vector<4xi1>, vector<4xi32> into vector<4xi32>
+  vector.print %v1 : vector<4xi32>
+
+
+  // expandload with multi-dimension memref
+  //    case 2: cross the most-inner dimension
+  // In this case, it will behavior like the memref is flat
```
Comment (zhanghb97) [inline examples/MLIRVector/vector-expandload.mlir:58]: behavior -> behave

File: examples/MLIRVector/vector-expandload.mlir [hunk 4]
```diff
@@ -0,0 +1,102 @@
...
+  %mask4 = arith.constant dense<[1, 0, 1, 1, 1, 1, 0, 0]> : vector<8xi1>
+
+  %v4 = vector.expandload %base4[%c1, %c1], %mask4, %pass_thru_8
+    : memref<?x?xi32>, vector<8xi1>, vector<8xi32> into vector<8xi32>
+  vector.print %v4 : vector<8xi32>
+
+
+  // expandload with dynamic memref
+  //    case 2: out-of-bound
+  // it behavior likes vector.load -- it's a platform-specified operation
```
Comment (zhanghb97) [inline examples/MLIRVector/vector-expandload.mlir:83]: behavior -> behaves

platform-specified -> platform-specific

File: examples/MLIRVector/vector-gather.mlir [hunk 5]
```diff
@@ -0,0 +1,82 @@
...
+                                                        [4, 5, 6, 7],
+                                                        [8, 9, 10, 11],
+                                                        [12, 13, 14, 15]]>
+
+memref.global "private" @gv2 : memref<8xi32> = dense<[0, 1, 2, 3, 4, 5, 6, 7]>
+
+func.func private @printMemrefI32(memref<*xi32>)
+
+func.func @main() -> i32 {
+  // vector.gather is also load with mask, but loading element in custom order,
```
Comment (zhanghb97) [inline examples/MLIRVector/vector-gather.mlir:13]: vector.gather also loads with a mask, but it loads elements in a custom order,

File: examples/MLIRVector/vector-gather.mlir [hunk 6]
```diff
@@ -0,0 +1,82 @@
...
+
+func.func private @printMemrefI32(memref<*xi32>)
+
+func.func @main() -> i32 {
+  // vector.gather is also load with mask, but loading element in custom order, 
+  // rather than sequentially:
+  //    result[0] := mask[0] ? base[index[0]] : pass_thru[0]
+  //    result[1] := mask[1] ? base[index[1]] : pass_thru[1]
+  //    ...
+  // As a comparison, that's what vector.maskedload doing:
```
Comment (zhanghb97) [inline examples/MLIRVector/vector-gather.mlir:18]: doing -> does

File: examples/MLIRVector/vector-gather.mlir [hunk 7]
```diff
@@ -0,0 +1,82 @@
...
+  //    result[0] := mask[0] ? base[0] : pass_thru[0]
+  //    result[1] := mask[1] ? base[1] : pass_thru[1]
+  //    ...
+  // But vector.gather support loading n-D vector.
```
Comment (zhanghb97) [inline examples/MLIRVector/vector-gather.mlir:22]: support -> supports

File: examples/MLIRVector/vector-load.mlir [hunk 8]
```diff
@@ -1,22 +1,89 @@
...
+
+  // load with m-D memref
+  //  case 1: inside inner-most dimension
+  %v1 = vector.load %base1[%c1, %c0] : memref<4x4xi32>, vector<4xi32>
+  vector.print %v1 : vector<4xi32>
+
+
+  // load with m-D memref
+  //  case 2 : cross inner-most dimension
+  // In this case, it will behavior like the memref is "flat"
```
Comment (zhanghb97) [inline examples/MLIRVector/vector-load.mlir:38]: behavior -> behave

File: examples/MLIRVector/vector-maskedload.mlir [hunk 9]
```diff
@@ -1,15 +1,88 @@
...
+                                                        [8, 9, 10, 11],
+                                                        [12, 13, 14, 15]]>
+memref.global "private" @gv2 : memref<8xi32> = dense<[0, 1, 2, 3, 4, 5, 6, 7]>
+
+func.func private @printMemrefI32(memref<*xi32>)
 
 func.func @main() -> i32 {
-  %base = memref.get_global @gv : memref<9xi32>
-  %mask = arith.constant dense<[1, 1, 1, 0, 1, 0, 0, 1, 1]> : vector<9xi1>
+  // maskedload is load with mask, supporting to load 1-D vector into n-D memref with mask.
```
Comment (zhanghb97) [inline examples/MLIRVector/vector-maskedload.mlir:12]: is load -> loads

File: examples/MLIRVector/vector-maskedload.mlir [hunk 10]
```diff
@@ -1,15 +1,88 @@
...
+
+  %base0 = memref.get_global @gv0 : memref<8xi32>
+  %base1 = memref.get_global @gv1 : memref<4x4xi32>
+  %base2 = memref.get_global @gv2 : memref<8xi32>
+
+  %pass_thru_4 = arith.constant dense<[2330, 2331, 2332, 2333]> : vector<4xi32>
+  %pass_thru_8 = arith.constant dense<[2330, 2331, 2332, 2333, 2334, 2335, 2336, 2337]> : vector<8xi32>
+
+  // maskedload normal usage
+  // maskedload require a pass through value at any time
```
Comment (zhanghb97) [inline examples/MLIRVector/vector-maskedload.mlir:28]: requires

File: examples/MLIRVector/vector-maskedload.mlir [hunk 11]
```diff
@@ -1,15 +1,88 @@
...
+  %mask1 = arith.constant dense<[1, 0, 0, 1]> : vector<4xi1>
+
+  %v1 = vector.maskedload %base1[%c0, %c0], %mask1, %pass_thru_4
+    : memref<4x4xi32>, vector<4xi1>, vector<4xi32> into vector<4xi32>
+  vector.print %v1 : vector<4xi32>
+
+
+  // maskedload with multi-dimension memref
+  //    case 2: cross the most-inner dimension
+  // In this case, it will behavior like the memref is flat
```
Comment (zhanghb97) [inline examples/MLIRVector/vector-maskedload.mlir:45]: behavior -> behave

File: examples/MLIRVector/vector-maskedload.mlir [hunk 12]
```diff
@@ -1,15 +1,88 @@
...
+  %mask4 = arith.constant dense<[1, 0, 1, 1, 1, 1, 0, 0]> : vector<8xi1>
+
+  %v4 = vector.maskedload %base4[%c1, %c1], %mask4, %pass_thru_8
+    : memref<?x?xi32>, vector<8xi1>, vector<8xi32> into vector<8xi32>
+  vector.print %v4 : vector<8xi32>
+
+
+  // maskedload with dynamic memref
+  //    case 2: out-of-bound
+  // it behavior likes vector.load -- it's a platform-specified operation
```
Comment (zhanghb97) [inline examples/MLIRVector/vector-maskedload.mlir:70]: behavior -> behaves
platform-specified -> platform-specific

File: examples/MLIRVector/vector-maskedstore.mlir [hunk 13]
```diff
@@ -1,32 +1,123 @@
...
+                                                        [12, 13, 14, 15]]>
+
+memref.global "private" @gv3 : memref<8xi32> = dense<[0, 1, 2, 3, 4, 5, 6, 7]>
+
 func.func private @printMemrefI32(memref<*xi32>)
 
 func.func @main() -> i32 {
-  %mask0 = arith.constant dense<[0, 1, 0, 1, 0, 1, 0, 1, 1]> : vector<9xi1>
-  %value0 = arith.constant dense<[12, 78, 54, 32, 34, 36, 39, 90, 21]> : vector<9xi32>
+  // maskedstore is store with mask, supporting store 1-D vector into n-D memref with mask.
```
Comment (zhanghb97) [inline examples/MLIRVector/vector-maskedstore.mlir:18]: is store -> stores

File: examples/MLIRVector/vector-maskedstore.mlir [hunk 14]
```diff
@@ -1,32 +1,123 @@
...
+
+  vector.maskedstore %base1[%c0, %c0], %mask1, %value1 
+    : memref<4x4xi32>, vector<4xi1>, vector<4xi32>
+
+  func.call @printMemrefI32(%gv1_for_print) : (memref<*xi32>) -> ()
+
+
+  // maskedstore with multi-dimension memref
+  //    case 2: cross the most-inner dimension
+  // In this case, it will behavior like the memref is flat
```
Comment (zhanghb97) [inline examples/MLIRVector/vector-maskedstore.mlir:60]: behavior -> behave

File: examples/MLIRVector/vector-maskedstore.mlir [hunk 15]
```diff
@@ -1,32 +1,123 @@
...
+
+  func.call @printMemrefI32(%gv1_for_print) : (memref<*xi32>) -> ()
+
+
+  // maskedstore with memref with custom layout
+  // TODO: find out how to create a memref with arbitrarily affine map layout
+  // "3" is reserved for this example
+
+  //============================================================================
+  // Tips: because keep using same memory region for all examples will make the
```
Comment (zhanghb97) [inline examples/MLIRVector/vector-maskedstore.mlir:75]: Remove "because"

File: examples/MLIRVector/vector-maskedstore.mlir [hunk 16]
```diff
@@ -1,32 +1,123 @@
...
+  %value4 = arith.constant dense<[400, 401, 402, 403, 404, 405, 406, 407]> : vector<8xi32>
+
+  vector.maskedstore %base4[%c1, %c1], %mask4, %value4
+    : memref<?x?xi32>, vector<8xi1>, vector<8xi32>
+
+  func.call @printMemrefI32(%gv2_for_print) : (memref<*xi32>) -> ()
+
+
+  // maskedstore with dynamic memref
+  //    case 2: what will happened if we store into somewhere out of bound?
```
Comment (zhanghb97) [inline examples/MLIRVector/vector-maskedstore.mlir:93]: happened -> happen

File: examples/MLIRVector/vector-scatter.mlir [hunk 17]
```diff
@@ -1,34 +1,66 @@
...
+memref.global "private" @gv0 : memref<8xi32> = dense<[0, 1, 2, 3, 4, 5, 6, 7]>
+
+memref.global "private" @gv1 : memref<8xi32> = dense<[0, 1, 2, 3, 4, 5, 6, 7]>
 
 func.func private @printMemrefI32(memref<*xi32>)
 
 func.func @main() -> i32 {
-  %cons1 = arith.constant 1 : index
-  %index0 = arith.constant dense<[1, 3, 7, 9, 0, 6, 5, 4, 2, 8]> : vector<10xi32> // index vector
+  // vector.scatter is also store with mask, but storing element in custom order,
```
Comment (zhanghb97) [inline examples/MLIRVector/vector-scatter.mlir:8]: vector.scatter also stores with mask, but it stores elements in a custom order,

File: examples/MLIRVector/vector-store.mlir [hunk 18]
```diff
@@ -0,0 +1,133 @@
...
+  // store with multi-dimension memref
+  //    case 1: inside most-inner dimension
+  %value1 = arith.constant dense<[200, 201, 202, 203]> : vector<4xi32>
+  vector.store %value1, %base1[%c0, %c0] : memref<4x4xi32>, vector<4xi32>
+  func.call @printMemrefI32(%gv1_for_print) : (memref<*xi32>) -> ()
+
+
+  // store with multi-dimension memref
+  //    case 2: cross the most-inner dimension
+  // In this case, it will behavior like the memref is flat
```
Comment (zhanghb97) [inline examples/MLIRVector/vector-store.mlir:52]: behavior -> behave

File: examples/MLIRVector/vector-store.mlir [hunk 19]
```diff
@@ -0,0 +1,133 @@
...
+  // vector.store %value4, %base3[%c2] : memref<8xi32>, vector<2x2xi32>
+  // func.call @printMemrefI32(%gv3_for_print) : (memref<*xi32>) -> ()
+  
+
+  // store with memref with custom layout
+  // TODO: find out how to create a memref with arbitrarily affine map layout
+  // "5" is reserved for this example
+
+  //============================================================================
+  // Tips: because keep using same memory region for all examples will make the
```
Comment (zhanghb97) [inline examples/MLIRVector/vector-store.mlir:90]: Remove "because"

File: examples/MLIRVector/makefile [hunk 20]
```diff
@@ -604,3 +626,24 @@ vector-contract-run:
...
+	${MLIR_TRANSLATE} --mlir-to-llvmir -o log.ll
+
+run-targets += vector-store-run
+vector-store-run:
+	@${MLIR_OPT} ./vector-store.mlir \
+		--convert-vector-to-scf --lower-affine --convert-scf-to-cf \
+		--convert-vector-to-llvm --convert-memref-to-llvm --convert-func-to-llvm \
+		--reconcile-unrealized-casts | \
+	${MLIR_CPU_RUNNER} ${OPT_FLAG} -e main -entry-point-result=i32 \
+		-shared-libs=${MLIR_RUNNER_UTILS} -shared-libs=${MLIR_C_RUNNER_UTILS}
```
Comment (zhanghb97) [inline examples/MLIRVector/makefile:649]: As the warning here, please add a newline at the end of the file.

File: examples/MLIRVector/vector-store.mlir [hunk 21]
```diff
@@ -0,0 +1,133 @@
...
+  %base8 = memref.cast %base2 : memref<4x4xi32> to memref<*xi32>
+  %mask8 = arith.constant dense<[1, 0, 0, 1]> : vector<4xi1>
+  %value8 = arith.constant dense<[800, 801, 802, 803]> : vector<4xi32>
+
+  // vector.maskedstore %base8[%c0, %c0], %mask8, %value8
+  //   : memref<*xi32>, vector<4xi1>, vector<4xi32>
+
+  %ret = arith.constant 0 : i32
+  return %ret : i32
+}
```
Comment (zhanghb97) [inline examples/MLIRVector/vector-store.mlir:133]: As the warning here, please add a newline at the end of the file.

**Review by zhanghb97 — CHANGES_REQUESTED** [review]

**Comment by Origami404** [comment]
Rebased onto main and fixed syntax errors mentioned here (and more by grammarly)

**Pull request merged**